### PR TITLE
New exchange: DigiFinex

### PR DESCRIPTION
**Pull request by szmcdull**

### CODE REVIEW

File: js/digifinex.js [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+        if (code === 0)
+            return;
+        caller = '[' + caller + '] ';
+        if (code === undefined)
+            throw new BadResponse (caller + 'bad response: ' + response);
+        this.checkCode (code, caller);
+    }
+
+    checkCode (code, caller) {
+        if (code === 10001) {
```
Comment (kroitor) [inline js/digifinex.js]: I suggest that we do the error handling in a unified way, as done in for, example:
- [liqui](https://github.com/ccxt/ccxt/blob/master/js/liqui.js#L76) + [`handleErrors()`](https://github.com/ccxt/ccxt/blob/master/js/liqui.js#L801)
- [bitfinex](https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L283) + [`handleErrors()`](https://github.com/ccxt/ccxt/blob/master/js/bitfinex.js#L1041)

File: js/digifinex.js [hunk 2]
```diff
@@ -0,0 +1,535 @@
...
+        return result;
+    }
+
+    dateUTC8 (timestampMS) {
+        const timedelta = this.safeValue (this.options, 'timedelta', 8 * 60 * 60 * 1000); // eight hours
+        return this.ymd (timestampMS + timedelta);
+    }
+
+    handleErrors (statusCode, statusText, url, method, responseHeaders, responseBody, response) {
+        this.checkResponse (response, '');
```
Comment (kroitor) [inline js/digifinex.js]: This is getting better, now I suggest that we don't have `handleErrors` method just for calling another method – instead the body of `checkResponse` should be in `handleErrors`, and I suggest that we rewrite `checkCode` to use an `.exceptions` map like it's done in liqui or bitfinex.

Comment (szmcdull) [reply js/digifinex.js]: I don't think .exceptions map is good here. A switch-like if-else group is more straight forward.

Comment (kroitor) [reply js/digifinex.js]: > A switch-like if-else group is more straight forward.

Nope, unfortunately, it's not, and it's much less convenient for the user – if a new exception code or message is added or changed – you could handle this via `.exceptions` without having to edit the code inside the method and rebuild the lib again. Basically the `.exceptions` mechanism is the standard way of doing it in CCXT, and if you see a different type of handling with if-else – that's mostly for historical reasons and to be replaced for `.exceptions`. So, in this case, and for most of new integrations, I suggest to stick to the unified approach, as it is done in other places throughout the lib. Most users will be really thankful for that. Let me know if you agree to the above )

File: js/digifinex.js [hunk 3]
```diff
@@ -0,0 +1,528 @@
+'use strict';
+
+//  ---------------------------------------------------------------------------
+
+const Exchange = require ('./base/Exchange');
+const { AccountSuspended, BadRequest, BadResponse, NetworkError, DDoSProtection, AuthenticationError, PermissionDenied, ArgumentsRequired, ExchangeError, InsufficientFunds, InvalidOrder, InvalidNonce } = require ('./base/errors');
+const { DECIMAL_PLACES } = require ('./base/functions/number');
+// const log = require ('ololog');
```
Comment (kroitor) [inline js/digifinex.js]: Wanted to ask you to also cleanup these commented debug logs throughout this file, pls )

File: js/digifinex.js [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+            'apiKey': '15cb6aad284ab8',
+            'secret': '48cad555befb5de238f6688b440a803f05cb6aad2',
+        });
+    }
+
+    async fetchMarkets (params = {}) {
+        let response = await this.privateGetTradePairs ();
+        // if (response.code != 0)
+        //     log.info ('[fetchMarkets] error: ' + response.code);
+        // console.log (response);
```
Comment (kroitor) [inline js/digifinex.js]: Let's cleanup these commented debug logs throughout this file )

File: js/digifinex.js [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+        let request = {
+            'symbol': symbol,
+        };
+        let response = await this.privateGetDepth (this.extend (request, params));
+        return this.parseOrderBook (response, parseInt (response['date']) * 1000);
+    }
+
+    async fetchTickers (symbols = undefined, params = {}) {
+        let response = await this.publicGetTicker ();
+        // log.bright.yellow.error (response);
```
Comment (kroitor) [inline js/digifinex.js]: Let's cleanup these commented debug logs throughout this file )

File: js/digifinex.js [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+        return result;
+    }
+
+    async cancelOrder (id, symbol = undefined, params = {}) {
+        // await this.loadMarkets ();
+        let response = await this.privatePostCancelOrder ({ 'order_id': id });
+        return response;
+    }
+
+    async CancelOrders (ids, symbol = undefined, params = {}) {
```
Comment (kroitor) [inline js/digifinex.js]: Need to rename this to camelcase (first character in lower case) – `CancelOrders` → `cancelOrders`

File: js/digifinex.js [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+        for (let i = 0; i < keys.length; i++) {
+            let uppercase = keys[i].toUpperCase ();
+            let account = this.account ();
+            if (this.inArray (keys[i], keysFree)) {
+                account['free'] = free[keys[i]];
+            }
+            if (this.inArray (keys[i], keysFrozen)) {
+                account['used'] = frozen[keys[i]];
+            }
+            account['total'] = account['free'] + account['used'];
```
Comment (kroitor) [inline js/digifinex.js]: We need to use `this.sum` in this line ↑ for reasons explained here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings

File: js/digifinex.js [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+    async fetchOHLCV (symbol, timeframe = '1m', since = undefined, limit = undefined, params = {}) {
+        let request = {
+            'symbol': symbol,
+            'type': 'kline_' + timeframe,
+        };
+        let response = await this.privateGetKline (this.extend (request, params));
+        return this.parseOHLCVs (response['data'], undefined, timeframe, since, limit);
+    }
+
+    nonce () {
```
Comment (kroitor) [inline js/digifinex.js]: We don't need to redefine the `nonce` method, if we are not calling it anywhere in this file.

File: js/digifinex.js (hunk 2 above)
Comment (szmcdull) [reply js/digifinex.js]: The error response of digifinex has only error code, without description message attached. If they add another error code, I must add a description myself. And it is more convenient to edit only the code, but not both the code and the map.

File: js/digifinex.js [hunk 9]
```diff
@@ -0,0 +1,528 @@
...
+                        'order_detail',
+                        'myposition',
+                    ],
+                    'post': [
+                        'trade',
+                        'cancel_order',
+                    ],
+                },
+            },
+            'commonCurrencies': {
```
Comment (kroitor) [inline js/digifinex.js]: This mapping should be removed, because we only need to have it if it converts one currency code to another currency code, in this case `commonCurrencies` does no conversions or unifications as [explained in the Manual](https://github.com/ccxt/ccxt/wiki/Manual#symbols-and-market-ids), so we can simply drop it from here.

File: js/digifinex.js [hunk 10]
```diff
@@ -0,0 +1,528 @@
...
+    async CancelOrders (ids, symbol = undefined, params = {}) {
+        // maximum 20 IDs supported
+        let response = await this.privatePostCancelOrder ({ 'order_id': ids.join (',') });
+        return response;
+    }
+
+    parseOrder (order, market = undefined) {
+        let side = order['type'];
+        let status = undefined;
+        if (order.status === 0 || order.status === 1) {
```
Comment (kroitor) [inline js/digifinex.js]: We need to switch the way these associative keys are accessed – to the single-quoted notation, because: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys

Comment (kroitor) [reply js/digifinex.js]: Also, see how `parseOrderStatus` is implemented in kraken. Let's do the same here, just copy it from kraken and change the string-numbers a bit.

File: js/digifinex.js (hunk 2 above)
Comment (szmcdull) [reply js/digifinex.js]: The error code is most likely not to change, only new codes will be added. And most users don't have to distinguish different exceptions at all.

Comment (kroitor) [reply js/digifinex.js]: > The error response of digifinex has only error code, without description message attached.

If the exchange does not provide an error message – I wouldn't add the text description in the first place we often return just the body of the response as is.

However, in some cases, if the user wants them, we have the mapping of exception codes to exception messages, and it's in the `describe()` as well, so if it comes to editing for a new message – the edit is still done in just one place (in the mapping inside `describe()`). We absolutely should not have the codes in one place and the messages in the other place. 

Here's what I'd ask you to do:

```JavaScript
    describe () {
        // ...
            'exceptions': {
                'exact': {
                    '10001': [ BadRequest, "Wrong request method, please check it's a GET ot POST request" ],
                    '10002': [ AuthenticationError, 'Invalid ApiKey' ],
                    '10003': [ AuthenticationError, "Sign doesn't match" ],
                    '10004': [ ArgumentsRequired, 'Illegal request parameters' ],
                    '10005': [ DDoSProtection, 'Request frequency exceeds the limit' ],
                    '10006': [ PermissionDenied, 'Unauthorized to execute this request' ],
                    '10007': [ PermissionDenied, 'IP address Unauthorized' ],
                    '10008': [ InvalidNonce, 'Timestamp for this request is invalid, timestamp must within 1 minute' ],
                    '10009': [ ExchangeError, 'Unexist endpoint, please check endpoint URL' ],
                    '10011': [ AccountSuspended, 'ApiKey expired. Please go to client side to re-create an ApiKey' ],
                    '20001': [ ExchangeError, 'Trade is not open for this trading pair' ],
                    '20002': [ ExchangeError, 'Trade of this trading pair is suspended' ],
                    '20003': [ InvalidOrder, 'Invalid price or amount' ],
                    '20007': [ InvalidOrder, 'Price precision error' ],
                    '20008': [ InvalidOrder, 'Amount precision error' ],
                    '20009': [ InvalidOrder, 'Amount is less than the minimum requirement' ],
                    '20010': [ InvalidOrder, 'Cash Amount is less than the minimum requirement' ],
                    '20011': [ InsufficientFunds, 'Insufficient balance' ],
                    '20012': [ BadRequest, 'Invalid trade type (valid value: buy/sell)' ],
                    '20013': [ OrderNotFound, 'No order info found' ],
                    '20014': [ BadRequest, 'Invalid date (Valid format: 2018-07-25)' ],
                    '20015': [ ExchangeError, 'Date exceeds the limit' ],
                    '20018': [ PermissionDenied, 'Your trading rights have been banned by the system' ],
                    '20019': [ BadRequest, 'Wrong trading pair symbol. Correct format:"usdt_btc". Quote asset is in the front' ],
                },
            },
        // ...
    }

    handleErrors (statusCode, statusText, url, method, responseHeaders, responseBody, response) {
        if (!response) {
            return; // fall back to default error handler
        }
        const code = this.safeString (response, 'code');
        if (code === '0') {
            return; // no error
        }
        if (code === undefined) {
            throw new BadResponse (this.id + ' ' + responseBody);
        }
        const feedback = this.id + ' ' + responseBody;
        const unknownError = [ ExchangeError, feedback ];
        const [ ExceptionClass, message ] = this.safeValue (this.exceptions['exact'], code, unknownError);
        throw new ExceptionClass (message);
    };
```

@szmcdull ↑ it's that easy and I think it is still better than having everything hardcoded, do you agree? )

Comment (kroitor) [reply js/digifinex.js]: > please check it's a GET ot POST request

↑ We usually don't copy their existing messages because their English sucks, btw.

Comment (szmcdull) [reply js/digifinex.js]: hahaha... OK. Can a key be a number, instead of a string?

Comment (kroitor) [reply js/digifinex.js]: > Can a key be a number, instead of a string?

It doesn't have to be anything but a string, and I've shown in my snippet above how the number-to-string conversion is done in `handleErrors()` in this line:

```JavaScript
        const code = this.safeString (response, 'code'); // safeString here instead of safeInteger
```

Let me know if that doesn't answer your question )

File: js/digifinex.js (hunk 9 above)
Comment (szmcdull) [reply js/digifinex.js]: Still not quite sure what it is for. I added DFT => DigiFinexToken. Is that alright?

Comment (kroitor) [reply js/digifinex.js]: > Still not quite sure what it is for

It is needed in cases where the exchange names the currency `XBT` meaning `BTC`. Or names it `DSH`, meaning `DASH`. Or in other such cases, where a renaming is needed.

However, for DFT – we don't need to rename it to DigiFinexToken, we would rather keep it as DFT everywhere.

So, in this particular case, the `.commonCurrencies` aren't needed at all and can be dropped.

**Comment by szmcdull** [comment]
Hi, @kroitor. Any update?

**Comment by szmcdull** [comment]
Hi @kroitor what should i do next? Why is this PR not accepted ha:)

**Comment by kroitor** [comment]
> Why is this PR not accepted ha:)

We need to clean it up before merging to make sure that all public and private unified structures correspond to https://github.com/ccxt/ccxt/wiki/Requirements, for that we need to fund the private accounts and run some tests with the exchange, that in itself requires time, and having a backlog of exchange integrations I don't always have the time to polish it. And if I merge it as is, other users will start copying bits of it all around, which we'd like to make clean first. 

Currently it has a few unification issues, the symbolic id vs symbols unification with `commonCurrencies` and `this.commonCurrencyCode` is not used throughout the file in the parsers, that is something I am going to rework as soon as I can.

Apart from that, we will need to make sure that all unified method signatures (`fetch*`, `parse*` and other) have the arguments standardized as listed in the base class or in the other exchange files.

There are also some typos like `statis` instead of `status`: https://github.com/ccxt/ccxt/pull/4993/files#diff-661bd8e86c285df2015cba2589f6f192R329

So, basically, we need to run the private set of tests with it and fix the issues there. If you want to run it locally, you will need to create a local `keys.local.json` file containing:

```JSON
{
    "digifinex": {
        "apiKey": "YOUR_API_KEY",
        "secret": "YOUR_SECRET"
    }
}
```

and then build and launch:

- `npm run build`
- `node js/test/test digifinex`

**Comment by kroitor** [comment]
I've added a few clarifications here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#new-exchange-integrations

Hope to get to it this particular file to polish it as soon as we can.

**Comment by szmcdull** [comment]
I ran the test and nothing seems failing.
```
$ node js/test/test digifinex

TESTING { exchange: "digifinex", symbol: "all" }  (<anonymous> @ test.js:33)

237 symbols [] (loadExchange @ test.js:176)
SYMBOL: btc_b2g (testExchange @ test.js:212)
fetching currencies not supported (exports [as fetchCurrencies] @ test.fetchCurrencies.js:35)
fetched all 237 tickers (exports [as fetchTickers] @ test.fetchTickers.js:35)
btc_b2g fetched 499 OHLCVs (exports [as fetchOHLCV] @ test.fetchOHLCV.js:21)
btc_b2g fetched 50 trades (exports [as fetchTrades] @ test.fetchTrades.js:22)
btc_b2g fetchOrderBook 2019-06-12T07:34:15.000Z bid: 0.00002314 bidVolume: 150 ask: 0.000024 askVolume: 786.4194 (printOrderBookOneLiner @ test.orderbook.js:18)
btc_b2g fetchL2OrderBook 2019-06-12T07:34:16.000Z bid: 0.00002314 bidVolume: 150 ask: 0.000024 askVolume: 786.4194 (printOrderBookOneLiner @ test.orderbook.js:18)
fetchOrderBooks() not supported (exports [as fetchOrderBooks] @ test.fetchOrderBooks.js:44)
fetching balance... (exports [as fetchBalance] @ test.fetchBalance.js:20)
zero balance (exports [as fetchBalance] @ test.fetchBalance.js:59)
fetching funding fees not supported (exports [as fetchFundingFees] @ test.fetchFundingFees.js:34)
fetching trading fees not supported (exports [as fetchTradingFees] @ test.fetchTradingFees.js:34)
fetching orders not supported (exports [as fetchOrders] @ test.fetchOrders.js:37)
fetched 0 open orders (exports [as fetchOpenOrders] @ test.fetchOpenOrders.js:24)
fetched 0 closed orders, testing each (exports [as fetchClosedOrders] @ test.fetchClosedOrders.js:22)
fetching my trades not supported (exports [as fetchMyTrades] @ test.fetchMyTrades.js:39)
fetching ledger items not supported (exports [as fetchLedger] @ test.fetchLedger.js:45)
```

I really appreciate your hard work to check all the details (I didn't realized the typo until you point it out:P). But somehow I'm still not very clear about the detailed specification of unification and standardization, what is wrong and why it is. Besides I cannot find the documentation for the internal APIs/methods (a manual for the users is available though). I have to figure out how things work from the code, but may still have no idea about commonCurrencies. Sometimes it is even not possible to know why it is like this or that until we ask:)

**Comment by kroitor** [comment]
> I ran the test and nothing seems failing.

```
zero balance (exports [as fetchBalance] @ test.fetchBalance.js:59)
fetching orders not supported (exports [as fetchOrders] @ test.fetchOrders.js:37)
fetched 0 open orders (exports [as fetchOpenOrders] @ test.fetchOpenOrders.js:24)
fetched 0 closed orders, testing each (exports [as fetchClosedOrders] @ test.fetchClosedOrders.js:22)
fetching my trades not supported (exports [as fetchMyTrades] @ test.fetchMyTrades.js:39)
```

↑ Despite that the tests didn't fail, there was no actual testing, because the major part of the private API formats that is critical – has not been tested (no data). The public API is ok, though (tested by travis without keys).

> But somehow I'm still not very clear about the detailed specification of unification and standardization, what is wrong and why it is. Besides I cannot find the documentation for the internal APIs/methods (a manual for the users is available though). 

Yes, I totally acknowledge this, the docs are in progress and are not complete in that regard. I will try to write up an integration guide for the devs, that will take the coder through the process. The docs for this project are also open-source, so we're editing them to reflect all the details as fast as we can (you're welcome as well, we appreciate all help), however, due to the backlog of requests, we don't always have a chance to do all the polishing work ourselves, due to reasons explained in the FAQ: https://github.com/ccxt/ccxt/wiki/FAQ

> I have to figure out how things work from the code, but may still have no idea about commonCurrencies. 

BTW, I've added a few clarifications on that to the Manual yesterday: https://github.com/ccxt/ccxt/blob/master/wiki/Manual.md#questions-on-naming-consistency

> Sometimes it is even not possible to know why it is like this or that until we ask:)

Sorry for that. I do understand that this is somewhat less convenient than it could be, yet it's still in the works, and we're trying to maintain it as much as we can. Apart from that it's a _"chicken vs egg"_ problem of a sort – the quality of the repo is evolving along with the requirements to it... Yes, in most cases it's better/safe to ask first, then do something, just to avoid duplication of work and common mistakes – will be happy to reply )

Thx!

**Comment by szmcdull** [comment]
> Despite that the tests didn't fail, there was no actual testing, because the major part of the private API formats that is critical – has not been tested (no data). The public API is ok, though (tested by travis without keys).

By private APIs you mean I need to test against an account with balance and trade history?

**Comment by kroitor** [comment]
> By private APIs you mean I need to test against an account with balance and trade history?

In general, yes, or, if you alternatively, you can take exchange responses from the docs and mockup the exchange responses in the exchange file itself temporarily and run the tests for the private formats, to make sure nothing is broken.

**Comment by nileio** [comment]
@kroitor @szmcdull appreciate your hardwork with this .. is this planned for merge with master or is there still checks to be completed !? thank you :)

**Comment by szmcdull** [comment]
@nileio I'll make some update recently, to fix some errors, and test against new APIs.

**Comment by szmcdull** [comment]
@kroitor Is there any test case for placing and canceling orders in the code base? How do you test trading?

**Comment by kroitor** [comment]
@szmcdull 

> Is there any test case for placing and canceling orders in the code base?
> How do you test trading?

We have test-cases for order exceptions here: 

- https://github.com/ccxt/ccxt/tree/master/js/test/errors

This is hard to design properly – because for extensive private tests you have to burn money. And It requires API keys, KYC and all of that, of course. So, in most cases we run fill tests locally. We have plans on making a robust system for testing PRs implementing private exchanges' APIs.

**Comment by szmcdull** [comment]
@kroitor I think we can, for example, place a bid order at extreme low price and examine if the order is place onto the order book.  Further more, I can contact digifinex to provide you an account for testing. I think they will be pleased to do so.

**Comment by kroitor** [comment]
@szmcdull 

> I think we can, for example, place a bid order at extreme low price and examine if the order is place onto the order book.

Of course, we can ) However, in order to do so we have to first place some money there. Very few exchanges actually have a working sandbox that would properly simulate the activity, so, it has to be done live with live money. Exchanges can provide an account for testing the trades with live money, of course, however, they will not provide accounts for testing, say, withdrawal capabilities and withdrawal history queries. When we test the private APIs and orders we should also test if the order does appear in the output of `fetchOpenOrders` and `fetchClosedOrders`. And to test if pagination actually works correctly – you have to make at least 100-1000 orders, deposits and withdrawals on average. Which is hard to do even if the exchange does provide an account for live-testing. You also have to use varying currencies and pairs in the tests, because most of the bugs don't appear on the most popular and the most active pairs. Even deciding which price would be "extreme low" for an inactive token in the absence of historical data – is a hard problem in itself. There are also limitations on precision, amounts and costs that must be accounted-for. So, I was mostly writing about generic private API testing, not about testing a specific account or a specific endpoint. 

> Further more, I can contact digifinex to provide you an account for testing. I think they will be pleased to do so.

I don't think there's a problem to sign up with DigiFinex and test the orders there, however, if you can contact them for a test-account – that would be very helpful! Thx a lot! 

The testing problem that we're trying to solve, however, is not with testing DigiFinex or any other specific exchange, because if it comes to testing a specific exchange most often I can still run the tests locally with my personal account. Therefore when we design generic tests, we have to consider the general case, that means unified testing of any exchange. It is hard to design and organize a testing infrastructure for an open-source community, because it requires a robust framework that would also keep the keys and the funds secure. We have some ideas on how to implement the tests that would satisfy our needs, however, we still have to work a bit more on those.

**Comment by szmcdull** [comment]
@kroitor I have a canceled order. Digifinex has fetchOpenOrders and fetchClosedOrders endpoints but not fetchOrders. The fetchClosedOrders endpoint returns this order but the test case complains that the status should be 'closed'. What is the correct way to handle this?

**Comment by kroitor** [comment]
> What is the correct way to handle this?

We should not lose the data. Many exchanges will return closed, canceled and partially-filled orders from the "inactive orders" endpoint, so, I think we should edit the test case to check for both `closed` and `canceled`.

File: js/digifinex.js [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+        for (let i = 0; i < keys.length; i++) {
+            const uppercase = keys[i].toUpperCase ();
+            const account = this.account ();
+            if (this.inArray (keys[i], keysFree)) {
+                account['free'] = free[keys[i]];
+            }
+            if (this.inArray (keys[i], keysFrozen)) {
+                account['used'] = frozen[keys[i]];
+            }
+            account['total'] = this.sum (account['free'] + account['used']);
```
Comment (kroitor) [inline js/digifinex.js]: Unfortunately, this is technically an error, it should be as described here: 

https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#adding-numbers-and-concatenating-strings

```JavaScript
// wrong
this.sum (account['free'] + account['used']);
// correct
this.sum (account['free'], account['used']);
```

However, since recently, we have changed and unified fetchBalance across all integrated exchanges, so, if you can copy the flow and style from https://github.com/ccxt/ccxt/blob/master/js/bittrex.js#L324 – that will be really great, namely we need to fix the following:

- use `safeCurrencyCode`
- drop `total` (it will be calculated by `parseBalance`)
- use `safeFloat` and `safeValue` to sanitize input as described here: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sanitizing-input-with-safe-methods

File: js/digifinex.js [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+            result[uppercase] = account;
+        }
+        return this.parseBalance (result);
+    }
+
+    async fetchOrderBook (symbol, limit = undefined, params = {}) {
+        // limit can be up to 100, defaults to 10
+        const markets = await this.loadMarkets ();
+        const market = markets[symbol];
+        symbol = market['baseId'] + '_' + market['quoteId'];
```
Comment (kroitor) [inline js/digifinex.js]: Unfortunately, this line is a violation of https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#sending-market-ids, your help with rewriting it for unification is really welcome!

File: js/digifinex.js [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+
+    async fetchTickers (symbols = undefined, params = {}) {
+        if (symbols !== undefined) {
+            throw new BadRequest ('Only supports fetching all tickers');
+        }
+        const response = await this.publicGetTicker ();
+        const result = {};
+        const keys = Object.keys (response['ticker']);
+        for (let i = 0; i < keys.length; i++) {
+            let symbol = keys[i].toUpperCase ();
```
Comment (kroitor) [inline js/digifinex.js]: In this case another symbol-related rule is volated: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#parsing-symbols. There are other similar lines throughout the file.

File: js/digifinex.js [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+        const request = {
+            'symbol': symbol,
+            'type': 'kline_' + timeframe,
+        };
+        const response = await this.privateGetKline (this.extend (request, params));
+        return this.parseOHLCVs (response['data'], market, timeframe, since, limit);
+    }
+
+    sign (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
+        params['apiKey'] = this.apiKey;
```
Comment (kroitor) [inline js/digifinex.js]: We should not mutate the argument passed from the outside in this case, need to rewrite this as shown, for example, here:
- https://github.com/ccxt/ccxt/blob/master/js/bitso.js#L283
- https://github.com/ccxt/ccxt/blob/master/js/deribit.js#L615

(`params` is always assigned to a different new object, but we don't change the originally-referenced object in the `params` argument)

File: js/digifinex.js (hunk 12 above)
Comment (szmcdull) [reply js/digifinex.js]: I tried follow your instruction. But their symbol id naming schema has changed in V3. The V3 order book API returns symbol id in the new format.

Comment (kroitor) [reply js/digifinex.js]: Oh, I see. No worries on that endpoint then.

Comment (szmcdull) [reply js/digifinex.js]: I tried follow your instruction. But their symbol id naming schema has changed in V3. The V3 order book API returns symbol id in the new format.

File: js/digifinex.js (hunk 11 above)
Comment (szmcdull) [reply js/digifinex.js]: The balance data format returned by Digifinex is really weird. Each currency may contains only the free field but not the used field. So the sum() computation cannot be omitted.

**Comment by kroitor** [comment]
@szmcdull i'll rework it for v3 and hope to merge soon.
